File: examples/FsharpExamples/CustomProps.fs
module CustomProps

// Learn more about F# at http://fsharp.org
open System
open Pulsar.Client.Api

open System.Text
open Pulsar.Client.Common
open FSharp.UMX

let runCustomProps () =

    let serviceUrl = "pulsar://my-pulsar-cluster:30002"
    let subscriptionName = "my-subscription"
    let topicName = sprintf "my-topic-%i" DateTime.Now.Ticks

    task {
        
        let! client =
            PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync()

        let! producer =
            client.NewProducer()
                .Topic(topicName)
                .CreateAsync()

        let! consumer =
            client.NewConsumer()
                .Topic(topicName)
                .SubscriptionName(subscriptionName)
                .SubscribeAsync()

        let payload = Encoding.UTF8.GetBytes(sprintf "Sent from F# at '%A'" DateTime.Now)
        let msg = producer.NewMessage(payload, "F#", readOnlyDict [("1","one")])
        let! messageId = producer.SendAsync(msg)
        printfn "MessageId is: '%A'" messageId

        let! message = consumer.ReceiveAsync()
        printfn "Received: %s key: %s prop1: %s" (message.Data |> Encoding.UTF8.GetString) %message.Key message.Properties.["1"]

        do! consumer.AcknowledgeAsync(message.MessageId)
    }

File: examples/CsharpExamples/CustomProps.cs
using System;
using System.Text;
using System.Threading.Tasks;
using Pulsar.Client.Api;
using System.Collections.Generic;

namespace CsharpExamples
{
    internal class CustomProps
    {
        internal static async Task RunCustomProps()
        {
            const string serviceUrl = "pulsar://my-pulsar-cluster:30002";
            const string subscriptionName = "my-subscription";
            var topicName = $"my-topic-{DateTime.Now.Ticks}";

            var client = await new PulsarClientBuilder()
                .ServiceUrl(serviceUrl)
                .BuildAsync();

            var producer = await client.NewProducer()
                .Topic(topicName)
                .CreateAsync();

            var consumer = await client.NewConsumer()
                .Topic(topicName)
                .SubscriptionName(subscriptionName)
                .SubscribeAsync();

            var payload = Encoding.UTF8.GetBytes($"Sent from C# at '{DateTime.Now}'");
            var msg = producer.NewMessage(payload, "C#", new Dictionary<string, string> { ["1"] = "one" });
            var messageId = await producer.SendAsync(msg);
            Console.WriteLine($"MessageId is: '{messageId}'");

            var message = await consumer.ReceiveAsync();
            Console.WriteLine($"Received: {Encoding.UTF8.GetString(message.Data)} key: {message.Key} prop1: {message.Properties["1"]}");

            await consumer.AcknowledgeAsync(message.MessageId);
        }
    }
}